### PR TITLE
Protein assign and crosslink site for loop and deadend

### PR DESCRIPTION
**Pull request by lonelu**
Solve #1676  and #1656

### CODE REVIEW

**Comment by lonelu** [comment]
Also solve #1581

File: EngineLayer/CrosslinkSearch/CrosslinkSpectralMatch.cs [hunk 1]
```diff
@@ -18,15 +18,46 @@ public CrosslinkSpectralMatch(PeptideWithSetModifications theBestPeptide, int no
         public CrosslinkSpectralMatch BetaPeptide { get; set; }
```
Comment (rmillikin) [inline EngineLayer/CrosslinkSearch/CrosslinkSpectralMatch.cs]: unit tests?

Comment (lonelu) [reply EngineLayer/CrosslinkSearch/CrosslinkSpectralMatch.cs]: This pull request is mainly on feature addition or bug fix. The unit tests are added based on existing unit tests.

Comment (rmillikin) [reply EngineLayer/CrosslinkSearch/CrosslinkSpectralMatch.cs]: please try to get >90% unit test coverage

File: TaskLayer/XLSearchTask/XLSearchTask.cs [hunk 2]
```diff
@@ -125,22 +130,61 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
 
             ReportProgress(new ProgressEventArgs(100, "Done with all searches!", new List<string> { taskId, "Individual Spectra Files" }));
 
-            var allPsmsList = allPsms.Select(p => p.First()).OrderByDescending(p => p.XLTotalScore).ToList();
+            List<List<CrosslinkSpectralMatch>> filteredAllPsms = new List<List<CrosslinkSpectralMatch>>();
 
-            foreach (var psm in allPsmsList)
+            foreach (var psms in allPsms)
```
[the remark this answers is not in the document]
Comment (lonelu) [reply TaskLayer/XLSearchTask/XLSearchTask.cs]: Name are changed to 'psmsPerScan'

File: TaskLayer/XLSearchTask/XLSearchTask.cs [hunk 3]
```diff
@@ -17,9 +17,14 @@ public class XLSearchTask : MetaMorpheusTask
     {
         public XLSearchTask() : base(MyTask.XLSearch)
         {
+            var digestPara = new DigestionParams(
```
[the remark this answers is not in the document]
Comment (lonelu) [reply TaskLayer/XLSearchTask/XLSearchTask.cs]: It is settable in GUI. This is just different from SearchTask.

File: EngineLayer/CrosslinkSearch/CrosslinkSpectralMatch.cs [hunk 4]
```diff
@@ -18,14 +18,69 @@ public CrosslinkSpectralMatch(PeptideWithSetModifications theBestPeptide, int no
...
+                            return true;
+                        }
+                    }
+
+                }
+            }
+            return false;
+        }
+
+        public void ResolveProteinPosAmbiguitiesForXl()
```
Comment (trishorts) [inline EngineLayer/CrosslinkSearch/CrosslinkSpectralMatch.cs]: all cases should have unit test

File: TaskLayer/XLSearchTask/XLSearchTask.cs [hunk 5]
```diff
@@ -125,22 +130,61 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
...
+                {
+                    continue;
+                }
+                string betaFulseq = "-";
+                if (csm.BetaPeptide != null)
+                {
+                    betaFulseq += csm.BetaPeptide.FullSequence;
+                }
+
+                if (keyValuePairs.ContainsKey(csm.FullSequence + betaFulseq))
```
Comment (trishorts) [inline TaskLayer/XLSearchTask/XLSearchTask.cs]: unit test

Comment (lonelu) [reply TaskLayer/XLSearchTask/XLSearchTask.cs]: Covered in unit test TestMixedMs2Ms2()''

**Pull request merged**